packages/gatsby/src/redux/nodes.ts

```diff
@@ -695,15 +695,16 @@ export const getNodesFromCacheByValue = (
     }
     const filterValueArr: Array<FilterValueNullable> = filterValue
 
-    const arr: Array<IGatsbyNode> = []
+    const set: Set<IGatsbyNode> = new Set()
 
     // TODO: we can also mergeSort for every step. this may perform worse because of how memory in js works.
     // For every value in the needle array, find the bucket of nodes for
     // that value, add this bucket of nodes to one list, return the list.
     filterValueArr.forEach((v: FilterValueNullable) =>
-      filterCache.byValue.get(v)?.forEach(v => arr.push(v))
+      filterCache.byValue.get(v)?.forEach(v => set.add(v))
     )
 
+    const arr = [...set] // this is bad for perf but will guarantee us a unique set :(
```

Comment:
would it help to have 1 set and 1 array? Memory will grow but maybe it's not a big deal? The set could contain only identifiers if that makes things smaller.

Reply:
I think V8 has a fast-path for cases like this, so shouldn't be that bad: https://v8.dev/blog/spread-elements#spreading-strings%2C-sets%2C-and-maps

Reply:
Beyond whether it can/should or not, concrete changes to the fast filter caching mechanism will take up more time than we currently have available. Either this is a band-aid that works for an operator that is already assumed to be slower than `$eq`, or it's the best fix and `$in` is still going to be slower. I would keep it like this for now.

I think there's still perf left ot be squeezed out of fast filters but that'll have to be done in like a focus sprint where I can get back into context of all the things. And after we finish the benchmark project, because then we can more reliably track its perf.

Reply:
(Once we have those graphs and the right benchmarks, we can easily compare different implementations against each other and make a decision with high degree of confidence based on numbers)